README.md
# GSGO Clip Generator
A system to cut extract kills form gameplay recordings using Global Offensive Game State Integration.

The system is baced around synchronising the recording from OBS with the data from the Game State Integration. It is designed to generate 1 second long clips with the kill half way through.
If OBS is setup as correctly the clips should be lossless and 60fps.


## Requirements:
* Python (https://www.python.org/downloads/)
* NodeJs (https://nodejs.org/en/download/)
* OBS    (https://obsproject.com/download)


# Setup:

1. OBS Settings (settings button bottem right)
	1. Set recording
		>_<Settings>Output>Recording Quality> _**Lossless Qualitym Tremendously Large File Size**
	3. Recording Path
		>_Settings>Output>Recording Path> **[Same location as RecordingSlicerVxxx.py]**_
	4. Record Hotkey
		>_Settings>Hotkeys>Start Recording: **[Start Recording KEY]**_ i.e. F8
	5. Record Hotkey
		>_Settings>Hotkeys>Start Stop Recording: **[Stop Recording KEY]**_ i.e. F9
	
	6. FPS
		>_Settings>Video>FPS Values:_ **60**

2. Global Offensive Game State Integration
	* Copy file:  
		**`gamestate_integration_yourservicenamehere.cfg`**  
		into  
		`\Steam\steamapps\common\Counter-Strike Global Offensive\csgo\cfg`

	i.e. steam libary location, mine is (`E:\Steam\steamapps\common\Counter-Strike Global Offensive\csgo\cfg`)


# Recording:

1. Run the CSGOStats.js applicaton using node.  
	- Option A. Start a command promt and navigate to the file (in it's current location), then run `node CSGOStats.js`  
	  
	- Optoin B. Right click the `CSGOStats.js` and _Open With > Choose another app> more apps >_ `"C:\Program Files\nodejs\node.exe"` **_(recommended)_**  
        _(Option B will allow you to run the node application in future by doubble clicking it.)_

2. Launch OBS, *(don't start recording yet)*
3. Join a CSGO session.
4. Tab to the node comand prompt window.
5. Once a stream of numbers starts, Press record using your hot key (i.e. f8) ***ENSURING THE TERMINAL IS IN THE RECORDING AREA!***
6. Play a round (first to 8). _(I recommend only playing one round as the recording is huge and will be 45gb+)_ but the system does now supports endless rounds.
7. Access the menu to trigger a save of the game data. _(Alt+F4 will **NOT** trigger save)_ (Do this atleast once before closing the game)
8. Close node window


# Convert to Clips:

If all has gone well you should have an .avi file with game play and a data.txt file of game data.

1. Rename the .avi file to `input.avi`

2. Run `RecordingSlicer.py`  
Follow the instructions:  
	- Enter the time at the bottom of the terminal in the extracted start image.  
	- Enter you in game name (same as in the data.txt file).  
	- You should be presented with a list of times in miliseconds.  

3. Press <kbd>ENTER</kbd> and your clips should appear in the `\output` folder.


**! I recommend checking each one as sometimes (such as when the bomb kill the person you are spectating) it incorrectly captuers.**  
**! I also recomend moving the files out of the output folder once you've checked them.**


Changelog.txt
V002
-stability and code fixing
-added support for endlesss rounds in one recording (README Recording 7.)
-obs setup fps increase form 30 to 60 (README Setup: 1. d.)

CSGOStats.js
http = require('http');
fs = require('fs');
var newdata = false;
serverport = 3005;
host = '127.0.0.1';
var newData = false;
var newFile = true;
var arrayData = [];
server = http.createServer( function(req, res)
{
    if (req.method == 'POST')
	{
        res.writeHead(200, {'Content-Type': 'text/html'});
        var body = '';
        req.on('data', function (data)
		{
            body += data;
        });
        req.on('end', function ()
		{
        	res.end( '' );

			if (res.statusCode === 200)
			{
				try
				{
					var data = JSON.parse(body);
          var activity = data.player.activity;
          if(activity == "playing"){
            var name = data.player.name;
            var num = parseInt(data.player.state.round_kills)+parseInt(data.player.state.round_killhs);
  					var ts = Date.now().toString().substring(4, 12);
  					//console.log(ts +", " + name + ",  "+ num);
            console.log(ts.substring(3) +", " + name + ",  "+ num);
  					arrayData.push([ts, name, num]);
  					newData = true;
          }
          else{
            if(newData){
              if(newFile){
                fs.writeFileSync('data.txt', arrayData.join("\n"));
                newFile = false;
              }
              else{
                fs.appendFileSync('data.txt', arrayData.join("\n"));
              }
              arrayData = [[]];
              console.log("Left game session | Saved session data");
              newData = false;
            }
            console.log("Not in Game...");
          }
				} catch (e)
				{
					console.log('Not geing valid Game session data');
				}
			}
        });
    }
    else
    {
        console.log("Not expecting other request types...");
        res.writeHead(200, {'Content-Type': 'text/html'});
		var html = '&lt;html>&lt;body>HTTP Server at http://' + host + ':' + serverport + '&lt;/body>&lt;/html>';
        res.end(html);
    }
});

server.listen(serverport, host);
console.log('Listening at http://' + host + ':' + serverport);


RecordingSlicerV002.py
import subprocess
import os

#videoFilename = str(input("Enter Gameplay Video file name (ensure it share the same directory as split.py): "))
#dataFilename = str(input("Enter Gameplay Data file name (ensure it share the same directory as split.py): "))
videoFilename = "input.avi"
dataFilename = "data.txt"


subprocess.call(["ffmpeg.exe", "-hide_banner", "-loglevel", "error", "-i", videoFilename, "-vf", "select=eq(n\\,1)", "-vframes", "1", "firstFrame.jpeg"])
subprocess.run(["explorer", "firstFrame.jpeg"])

kills = []

#print(subprocess.run(["stat",videoFilename]))
#wmic datafile where name=C:\\Users\\Matt\\Desktop\\DemoGameplay\\Clip Generator\\input.avi get creationdate | findstr /brc:[0-9]
fileTime = round(float(os.path.getctime(videoFilename)),3)*1000
print(str(fileTime)[7:-3])
fileTimeStartDigits = str(fileTime)[4:-7]
timeOffset = int(fileTimeStartDigits + str(input("Enter time offset (time at the bottem of the terminal): ")))


name = str(input("Enter In-Game Name: "))
#name = ""

os.remove("firstFrame.jpeg")

f = open(dataFilename, "r", encoding="utf8")
data = f.read()
lines = data.split("\n")
k=0

for lineNum in range(0,len(lines)):
    a = lines[lineNum].split(",")
    if(a[1] == name):
        if(k < int(a[2])):
            k = int(a[2])
            a[0] = int(a[0]) - timeOffset
            kills += [a[0]]
            print(a)
    else:
        k = 0
input(">")
for killTime in kills:
    outputFilename = "output\\" + str(killTime + timeOffset) + "." + videoFilename.split(".")[1]
    startTime = str((killTime/100)-0.5)
    subprocess.run(["ffmpeg.exe", "-hide_banner", "-loglevel", "error", "-ss", startTime, "-i", videoFilename, "-c", "copy", "-t", "1", outputFilename])
input(">")


CSGOStatsV002.js
http = require('http');
fs = require('fs');
var newdata = false;
serverport = 3005;
host = '127.0.0.1';
var newData = false;
var newFile = true;
var arrayData = [];
server = http.createServer( function(req, res)
{
    if (req.method == 'POST')
	{
        res.writeHead(200, {'Content-Type': 'text/html'});
        var body = '';
        req.on('data', function (data)
		{
            body += data;
        });
        req.on('end', function ()
		{
        	res.end( '' );

			if (res.statusCode === 200)
			{
				try
				{
					var data = JSON.parse(body);
          var activity = data.player.activity;
          if(activity == "playing"){
            var name = data.player.name;
            var num = parseInt(data.player.state.round_kills)+parseInt(data.player.state.round_killhs);
  					var ts = Date.now().toString().substring(4, 12);
  					//console.log(ts +", " + name + ",  "+ num);
            console.log(ts.substring(4) +", " + name + ",  "+ num);
  					arrayData.push([ts, name, num]);
  					newData = true;
          }
          else{
            if(newData){
              if(newFile){
                fs.writeFileSync('data.txt', arrayData.join("\n"));
                newFile = false;
              }
              else{
                fs.appendFileSync('data.txt', arrayData.join("\n"));
              }
              arrayData = [[]];
              console.log("Left game session | Saved session data");
              newData = false;
            }
            console.log("Not in Game...");
          }
				} catch (e)
				{
					console.log('Not geing valid Game session data');
				}
			}
        });
    }
    else
    {
        console.log("Not expecting other request types...");
        res.writeHead(200, {'Content-Type': 'text/html'});
		var html = '&lt;html>&lt;body>HTTP Server at http://' + host + ':' + serverport + '&lt;/body>&lt;/html>';
        res.end(html);
    }
});

server.listen(serverport, host);
console.log('Listening at http://' + host + ':' + serverport);
